src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/Transmission/OtlpExporterPersistentStorageTransmissionHandler.cs

```diff
@@ -0,0 +1,175 @@
+        {
+            Name = $"OtlpExporter Persistent Retry Storage - {typeof(TRequest)}",
+            IsBackground = true,
+        };
+
+        this.thread.Start();
+    }
+
+    // Used for test.
+    internal bool Forceflush(int timeOutMilliseconds)
+    {
+        this.exportEvent.Set();
+
+        if (this.dataExportNotification.WaitOne(timeOutMilliseconds))
+        {
+            return true;
+        }
+
+        return false;
+    }
+
+    protected override bool OnSubmitRequestFailure(TRequest request, ExportClientResponse response)
+    {
+        if (RetryHelper.ShouldRetryRequest(request, response, OtlpRetry.InitialBackoffMilliseconds, out _))
+        {
+            byte[]? data = null;
+            if (request is ExportTraceServiceRequest traceRequest)
+            {
+                data = traceRequest.ToByteArray();
+            }
+            else if (request is ExportMetricsServiceRequest metricsRequest)
+            {
+                data = metricsRequest.ToByteArray();
+            }
+            else if (request is ExportLogsServiceRequest logsRequest)
+            {
+                data = logsRequest.ToByteArray();
+            }
+
+            if (data != null)
+            {
+                this.persistentBlobProvider?.TryCreateBlob(data, out _);
+            }
+
+            return true;
+        }
+
+        return false;
+    }
+
+    protected override void OnShutdown(int timeoutMilliseconds)
+    {
+        try
+        {
+            this.shutdownEvent.Set();
+        }
+        catch (ObjectDisposedException)
+        {
+            // Dispose was called before shutdown.
+        }
+
+        this.thread.Join(timeoutMilliseconds);
+        base.OnShutdown(timeoutMilliseconds);
+    }
+
+    protected override void Dispose(bool disposing)
+    {
+        if (!this.disposed)
+        {
+            if (disposing)
+            {
+                this.shutdownEvent.Dispose();
+                this.exportEvent.Dispose();
+                this.dataExportNotification.Dispose();
+                (this.persistentBlobProvider as FileBlobProvider)?.Dispose();
+            }
+
+            this.disposed = true;
+        }
+    }
+
+    private void RetryStoredRequests()
+    {
+        while (true)
+        {
+            try
+            {
+                if (this.shutdownEvent.WaitOne(0))
+                {
+                    break;
+                }
+
+                // Wait 60 seconds before retrying
+                this.exportEvent.WaitOne(this.RetryIntervalInMilliseconds);
+
+                int fileCount = 0;
+
+                // TODO: Run maintenance job.
+                // Transmit 10 files at a time.
+                while (fileCount < 10 && !this.shutdownEvent.WaitOne(0))
+                {
+                    if (this.persistentBlobProvider.TryGetBlob(out var blob))
+                    {
+                        if (blob.TryLease((int)this.TimeoutMilliseconds) && blob.TryRead(out var data))
```

Comment:
* If any of these conditions fail we don't do any work but still perform `fileCount++;` issue or no?

* This `if (this.requestFactory != null)` seems to smell. Why would we not require a factory? Also the ctor doesn't allow `null` value but the property does. My guess is the issue here is that property is incorrectly annotated. I'll drop a couple suggestions to fix that.

Reply:
>If any of these conditions fail we don't do any work but still perform fileCount++; issue or no

We continue

> This if (this.requestFactory != null) seems to smell. Why would we not require a factory? Also the ctor doesn't allow null value but the property does. My guess is the issue here is that property is incorrectly annotated. I'll drop a couple suggestions to fix that.

Yes, fixed.